src/mapgen/cavegen.cpp

```diff
@@ -96,13 +101,30 @@ void CavesNoiseIntersection::generateCaves(MMVManip *vm,
 		u16 base_filler = depth_top + biome->depth_filler;
 		u16 depth_riverbed = biome->depth_riverbed;
 		u16 nplaced = 0;
+
+		cur_biome_depth = 0;
+		s16 nextBiomeY = biome_transitions[cur_biome_depth];
+
 		// Don't excavate the overgenerated stone at nmax.Y + 1,
 		// this creates a 'roof' over the tunnel, preventing light in
 		// tunnels at mapchunk borders when generating mapchunks upwards.
 		// This 'roof' is removed when the mapchunk above is generated.
 		for (s16 y = nmax.Y; y >= nmin.Y - 1; y--,
 				index3d -= m_ystride,
 				VoxelArea::add_y(em, vi, -1)) {
+			// We need this check to make sure that biomes don't generate too far down
+			if (y < nextBiomeY) {
+				biome = m_bmgn->getBiomeAtIndex(index2d, v3s16(x, y, z));
+				cur_biome_depth++;
+				nextBiomeY = biome_transitions[cur_biome_depth];
```

Comment:
Aren't these two lines redundant? The loop below should already increment it correctly.

Reply:
I moved the loop to up above before it starts looping through y values. I think it may be more efficient to not have the while loop run on every iteration. Once we know that we are starting at the correct position in the biome_transitions array then we can just go to the next value and know it will be correct.

Maybe it is just a micro optimization on branch prediction, but I figured I'd keep it since this code is going to be run a good amount.

Reply:
No sorry, there must be some misunderstanding. I meant to remove these two lines in particular and keep the loop where it is. To my understanding, it would then do one iteration at minimum anyway due to its (initially) true `y < biome_y_min` condition.